File: Leetcode_POTD_2025/7_JULY/898_Bitwise_ORs_of_Subarrays.cpp
class Solution {
public:
    int subarrayBitwiseORs(vector<int>& arr) {
        unordered_set<int> unique_ors;
        unordered_set<int> ending_ors;
        for (int x : arr) {
            unordered_set<int> next_ending_ors;
            next_ending_ors.insert(x);
            for (int y : ending_ors) {
                next_ending_ors.insert(x | y);
            }
            ending_ors = next_ending_ors;
            for (int val : ending_ors) {
                unique_ors.insert(val);
            }
        }
        return unique_ors.size();
    }
};

File: Leetcode_POTD_2025/7_JULY/898_Bitwise_ORs_of_Subarrays.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public int subarrayBitwiseORs(int[] arr) {
        Set<Integer> uniqueOrs = new HashSet<>();
        Set<Integer> currentOrs = new HashSet<>();
        for (int x : arr) {
            Set<Integer> nextOrs = new HashSet<>();
            nextOrs.add(x);
            for (int y : currentOrs) {
                nextOrs.add(x | y);
            }
            currentOrs = nextOrs;
            uniqueOrs.addAll(currentOrs);
        }
        return uniqueOrs.size();
    }
}
